### PR TITLE
Virtual keyboard

### PR DESCRIPTION
**Pull request by AnhelinaVr**
Ссылка на задание - https://github.com/rolling-scopes-school/tasks/blob/master/tasks/codejam-virtual-keyboard.md
Скриншот - https://prnt.sc/rv01es
Ссылка на демо- https://anhelinavr.github.io/Virtual-Keyboard/
Дата сдачи: 07.04.2020
Дата дедлайна: 07.04.2020
Самопроверка:
минимальный набор: реализован
стандартный набор: отсутствует сохранение выбранного языка. Остальное реализовано
дополнительный набор: реализован

### CODE REVIEW

File: style.css [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    }
+    
+    .keyboard-key[data-code='Enter'] {
+        width: 120px;
+    }
+    
+    .keyboard-key[data-code='ShiftLeft'],
+    .keyboard-key[data-code='ShiftRight'] {
+        width: 120px;
+    }
```
Comment (GattyGamaun) [inline style.css:72]: You can join selectors with the same properties.

File: src/script.js [hunk 2]
```diff
@@ -0,0 +1,843 @@
...
+        shift: {
+            isCaps: { en: '~', ru: 'ё' },
+            notCaps: { en: '~', ru: 'Ё' },
+        },
+        code: 'Backquote',
+    },
+    {
+        print: true,
+        text: { en: '1', ru: '1' },
+        CapsLock: { en: '1', ru: '1' },
```
Comment (GattyGamaun) [inline src/script.js:14]: PascalCase for properties!

File: src/script.js [hunk 3]
```diff
@@ -0,0 +1,843 @@
...
+        CapsLock: { en: 'Ctrl', ru: 'Ctrl' },
+        shift: {
+            isCaps: { en: 'Ctrl', ru: 'Ctrl' },
+            notCaps: { en: 'Ctrl', ru: 'Ctrl' },
+        },
+        code: 'ControlRight',
+    },
+];
+// create html elements
+const wrapper = document.createElement('DIV');
```
Comment (GattyGamaun) [inline src/script.js:585]: Use lowercase for parameter.

File: src/script.js [hunk 4]
```diff
@@ -0,0 +1,843 @@
...
+wrapper.appendChild(title);
+title.classList.add('title');
+title.textContent += 'Virtual Keyboard';
+const textarea = document.createElement('TEXTAREA');
+wrapper.appendChild(textarea);
+textarea.classList.add('screen');
+textarea.cols = 50;
+textarea.rows = 5;
+const keyboard = document.createElement('DIV');
+keyboard.id = 'keyboard';
```
Comment (GattyGamaun) [inline src/script.js:598]: Why do you need id here?

File: src/script.js [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+wrapper.appendChild(keyboard);
+const description = document.createElement('P');
+description.classList.add('description');
+description.innerHTML += 'Смена раскладки: Ctrl + Alt <br> Клавиатура создана в операционной системе Windows';
+wrapper.appendChild(description);
+
+let isCaps = false;
```
Comment (GattyGamaun) [inline src/script.js:605]: Use `const` instead of `let` were it possible.

File: src/script.js [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+    keyboardDraw(langEn);
+};
+
+
+textarea.onblur = () => {
+    textarea.focus();
+};
+
+function CapsLockActive(isCaps) {
+    isCaps ? document.querySelector('#keyboard .keyboard-key[data-code="CapsLock"]').classList.add('cl_active') : document.querySelector('#keyboard .keyboard-key[data-code="CapsLock"]').classList.remove('cl_active');
```
Comment (GattyGamaun) [inline src/script.js:620]: split it into a few lines for more readability

File: src/script.js [hunk 7]
```diff
@@ -0,0 +1,843 @@
+const Keyboard = [{
```
Comment (GattyGamaun) [inline src/script.js:1]: Use PascalCase only when naming constructors or classes

File: src/script.js [hunk 8]
```diff
@@ -0,0 +1,843 @@
...
+    langEn = localStorage.getItem('langEn');
+    keyboardDraw(langEn);
+};
+
+
+textarea.onblur = () => {
+    textarea.focus();
+};
+
+function CapsLockActive(isCaps) {
```
Comment (GattyGamaun) [inline src/script.js:619]: Use PascalCase only when naming constructors or classes

File: src/script.js [hunk 9]
```diff
@@ -0,0 +1,843 @@
...
+
+function keyboardDraw(langEn) {
+    let out = '';
+    for (let i = 0; i < Keyboard.length; i++) {
+        if (i === 14 || i === 29 || i === 42 || i === 55) out += '<div class="clear"></div>';
+        if (langEn === true) {
+            if ((i >= 0 && i < 12) || (i > 14 && i < 27) || (i > 29 && i < 41) || (i > 42 && i < 53)) out += `<div class="keyboard-key" onclick="keyClick(event)" data-code="${Keyboard[i].code}">${Keyboard[i].text.en}${Keyboard[i].shift.isCaps.ru.sub()}</div>`;
+            else out += `<div class="keyboard-key" onclick="keyClick(event)" data-code="${Keyboard[i].code}">${Keyboard[i].text.en}</div>`;
+        } else if ((i >= 0 && i < 12) || (i > 14 && i < 27) || (i > 29 && i < 41) || (i > 42 && i < 53)) out += `<div class="keyboard-key" onclick="keyClick(event)" data-code="${Keyboard[i].code}">${Keyboard[i].text.ru}${Keyboard[i].shift.isCaps.en.sub()}</div>`;
+        else out += `<div class="keyboard-key" onclick="keyClick(event)" data-code="${Keyboard[i].code}">${Keyboard[i].text.ru}</div>`;
```
Comment (GattyGamaun) [inline src/script.js:658]: Split it into multiple lines.

File: src/script.js [hunk 10]
```diff
@@ -0,0 +1,843 @@
...
+        keyboard.innerHTML = out;
+    }
+    CapsLockActive(isCaps);
+}
+
+
+function keyboardDrawCAPSLOCK(langEn) {
```
Comment (GattyGamaun) [inline src/script.js:665]: camelCase notation

File: src/script.js [hunk 11]
```diff
@@ -0,0 +1,843 @@
...
+        keyHandler(event);
+    } else if (Keyboard.find((elem) => elem.code === event.code).print === true || event.code === 'Backspace' || event.code === 'Delete' || event.code === 'ArrowLeft' || event.code === 'ArrowRight' || event.code === 'ArrowUp' || event.code === 'ArrowDown') {
+        keyHandler(event);
+    } else return;
+};
+
+document.onkeyup = function(event) {
+    document.querySelector(`#keyboard .keyboard-key[data-code="${event.code}"]`).classList.remove('active');
+
+    if (event.code == 'ShiftLeft' || event.code === 'ShiftRight') {
```
Comment (GattyGamaun) [inline src/script.js:823]: Use strict equality were it possible.

File: src/script.js [hunk 12]
```diff
@@ -0,0 +1,843 @@
...
+        isShift = false;
+        isCaps ? keyboardDrawCAPSLOCK(langEn) : keyboardDraw(langEn);
+        document.querySelector(`#keyboard .keyboard-key[data-code="${event.code}"]`).classList.remove('cl_active');
+    }
+};
+
+function keyClick(event) {
+    // console.log(event.target);
```
Comment (GattyGamaun) [inline src/script.js:831]: Comments code are not allowed.

File: style.css [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        text-align: center;
+        width: 100%;
+        font-size: 20px;
+    }
+    
+    .description {
+        font-size: 30px;
+    }
+    
+    #keyboard {
```
[the remark this answers is not in the document]
Comment (AnhelinaVr) [reply style.css:21]: I agree with your comment, it will be corrected. For CSS I needed to use classes.

File: style.css (hunk 1 above)
Comment (AnhelinaVr) [reply style.css:72]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 2 above)
Comment (AnhelinaVr) [reply src/script.js:14]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 3 above)
Comment (AnhelinaVr) [reply src/script.js:585]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 4 above)
Comment (AnhelinaVr) [reply src/script.js:598]: I agree with your comment, it will be corrected. For CSS I needed to use classes.

File: src/script.js (hunk 5 above)
Comment (AnhelinaVr) [reply src/script.js:605]: We cannot modify variables that are declared using `const` keyword. These variables change their content.

File: src/script.js (hunk 6 above)
Comment (AnhelinaVr) [reply src/script.js:620]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 7 above)
Comment (AnhelinaVr) [reply src/script.js:1]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 8 above)
Comment (AnhelinaVr) [reply src/script.js:619]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 9 above)
Comment (AnhelinaVr) [reply src/script.js:658]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 10 above)
Comment (AnhelinaVr) [reply src/script.js:665]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 11 above)
Comment (AnhelinaVr) [reply src/script.js:823]: I agree with your comment, it will be corrected.

File: src/script.js (hunk 12 above)
Comment (AnhelinaVr) [reply src/script.js:831]: I agree with your comment, it will be corrected.

**Comment by GattyGamaun** [comment]
20+10+8+15+15+8+10+10 = 96

**Pull request merged**